### PR TITLE
feature/category

### PR DESCRIPTION
**Pull request by bodya4243**
1) add category entity

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    public void deleteCategory(@PathVariable Long id) {
+        categoryService.deleteById(id);
+    }
+
+    @GetMapping("/{id}/books")
+    @PreAuthorize("hasAnyRole('ROLE_USER','ROLE_ADMIN')")
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@PathVariable Long id) {
+        return bookService.getBooksByCategoryId(id);
+    }
+
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/controller/CategoryController.java]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/dto/CategoryRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import jakarta.validation.constraints.Size;
+import lombok.Data;
+
+@Data
+public class CategoryRequestDto {
+    @NotNull
+    @Size(min = 2)
+    private String name;
+
+    private String description;
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/dto/CategoryRequestDto.java]: Please set max constraint to align with db constraints

File: src/main/java/com/example/onlinebookstore/mapper/BookMapper.java [hunk 3]
```diff
@@ -1,14 +1,31 @@
...
+
+    @AfterMapping
+    default void setCategoryIds(@MappingTarget BookResponseDto bookResponseDto, Book book) {
+        if (book.getCategoryId() != null) {
+            bookResponseDto.setCategoryIds(book.getCategoryId().stream()
+                    .map(Category::getId)
+                    .collect(Collectors.toSet()));
+        }
+    }
+
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/mapper/BookMapper.java:29]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/model/Book.java [hunk 4]
```diff
@@ -40,4 +46,12 @@ public class Book {
     
     @Column(nullable = false)
     private boolean isDeleted = false;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/model/Book.java]:
```suggestion
    @ManyToMany(fetch = FetchType.LAZY)
```

File: src/main/java/com/example/onlinebookstore/model/Book.java [hunk 5]
```diff
@@ -40,4 +46,12 @@ public class Book {
...
+    @JoinTable(
+            name = "books_categories",
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "category_id")
+    )
+    private Set<Category> categoryId = new HashSet<>();
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/model/Book.java:59]: Please exclude this from equals and hashCode

File: src/main/java/com/example/onlinebookstore/repository/BookRepository.java [hunk 6]
```diff
@@ -1,7 +1,9 @@
 package com.example.onlinebookstore.repository;
 
 import com.example.onlinebookstore.model.Book;
+import java.util.List;
 import org.springframework.data.jpa.repository.JpaRepository;
 
 public interface BookRepository extends JpaRepository<Book, Long> {
+    List<Book> findAllByCategoryId_Id(Long categoryId);
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/repository/BookRepository.java]: Better to keep naming convention

File: src/main/java/com/example/onlinebookstore/service/BookServiceImpl.java [hunk 7]
```diff
@@ -18,20 +19,20 @@ public class BookServiceImpl implements BookService {
     private final BookMapper bookMapper;
     
     @Override
-    public BookDto save(BookRequestDto bookDto) {
+    public BookResponseDto save(BookRequestDto bookDto) {
         Book book = bookMapper.toModel(bookDto);
         return bookMapper.toDto(bookRepository.save(book));
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/service/BookServiceImpl.java:24]: You should save and validate categories as well

File: src/main/resources/db/changelog/changes/insert-categories.yml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+            columns:
+              - column:
+                  name: name
+                  value: Fantasy
+              - column:
+                  name: description
+                  value: Fantasy books
+              - column:
+                  name: is_deleted
+                  valueBoolean: false
```
Comment (andrii-hoienko) [inline src/main/resources/db/changelog/changes/insert-categories.yml]:
```suggestion
                  valueBoolean: false
                  
```

File: src/main/resources/db/changelog/changes/insert-books-categories.yml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+                  value: 2
+        - insert:
+            tableName: books_categories
+            columns:
+              - column:
+                  name: book_id
+                  value: 3
+              - column:
+                  name: category_id
+                  value: 3
```
Comment (andrii-hoienko) [inline src/main/resources/db/changelog/changes/insert-books-categories.yml:32]: Please use valueComputed instead hardcoding id`s

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job! Left some comments

File: src/main/java/com/example/onlinebookstore/repository/BookRepository.java [hunk 10]
```diff
@@ -5,5 +5,5 @@
 import org.springframework.data.jpa.repository.JpaRepository;
 
 public interface BookRepository extends JpaRepository<Book, Long> {
-    List<Book> findAllByCategoryId_Id(Long categoryId);
+    List<Book> findAllById(Long categoryId);
```
Comment (yevheniisynytsia) [inline src/main/java/com/example/onlinebookstore/repository/BookRepository.java]: findAllById(id) in book repository sounds like we are looking for all books by **book** id. Please return categoryId back

There are two options:
1. Use ugly naming
2. Use nice naming but write a customer query

File: src/main/java/com/example/onlinebookstore/model/Category.java [hunk 11]
```diff
@@ -29,4 +29,12 @@ public class Category {
...
     @Column(nullable = false)
     private boolean isDeleted = false;
+
+    public Category(Long id) {
+        this.id = id;
+    }
+
+    public Category() {
+
+    }
```
Comment (yevheniisynytsia) [inline src/main/java/com/example/onlinebookstore/model/Category.java]: replace with `@NoArgsConstruuctor`

File: src/main/java/com/example/onlinebookstore/dto/BookDtoWithoutCategoryIds.java [hunk 12]
```diff
@@ -4,18 +4,18 @@
 import lombok.Data;
 
 @Data
-public class BookDto {
+public class BookDtoWithoutCategoryIds {
     private Long id;
-    
+
```
Comment (yevheniisynytsia) [inline src/main/java/com/example/onlinebookstore/dto/BookDtoWithoutCategoryIds.java]: please remove empty lines

Comment (yevheniisynytsia) [reply src/main/java/com/example/onlinebookstore/dto/BookDtoWithoutCategoryIds.java]: if there are a lot of annotations - it is ok to add empty lines, but when there are no annotations above the fields, I don't see a reason to have so much empty space

**Review by yevheniisynytsia — CHANGES_REQUESTED** [review]
Well done! Left some minor comments

**Pull request merged**